### DISCUSSION TITLE
AR Not Working

### DISCUSSION THREAD

**Discussion by ntxmodel** [google/model-viewer#4789] [General]
I am trying to develop my first web page to show a glb model in 3D with the ability to rotate, pan, and zoom on my laptop (I do not currently have access to a web site). I have used the `<script type="module" src="https://ajax.googleapis.com/ajax/libs/model-viewer/3.5.0/model-viewer.min.js"></script>` and the model shows up fine, but I cannot rotate, pan and zoom the model. My assumption is that I am really not getting access to threejs to support the AR features. I went to the Model-Viewer website and uploaded my glb file and then downloaded the code and that is what I am running except I added the script tag I showed.

I have tried to go through the documentation several times, but I am obviously doing something wrong. I am also curious if the editor changes the glb file and that I need to download the "modified" version (which is what I did do).

**Comment by danilokrulj** [comment]
Maybe sharing your code could help someone spot the issue? 
So the model loads for preview but you can't rotate, pan or zoom? And you can't start the AR session?

**Comment by ntxmodel** [comment]


**Reply by ntxmodel** [reply]
Yes, the model shows up but, you cannot rotate, pan or zoom.

I tried a couple of posts to show the code in a message but it did not work correctly so here is a txt version of the page to show the model. So, feel free to look at the text and let me know what I am doing wrong.

[index.txt](https://github.com/google/model-viewer/files/15437557/index.txt)

**Reply by vehrmann** [reply]
I guess you are seeing the poster, not the model.

Due to some security reasons, Javascript is not executed when you open html-files which are stored on your disk and not on a web-server.
For developping, you can use e.g. Visual Studio Code, then install the extension "Live Server" by publisher Ritwick Dey. This allows you to run a local live server which lets you execute your html/js correctly.

**Reply by ntxmodel** [reply]
Are you then required to use Visual Studio Code to do your coding (not that I am not willing to give it a try)?

Does this mean that I can start Live Server and then run any pages I want to? 

I have a "set" of web pages and the problem I have been experiencing is a couple of layers down from the "start page". If I start with the "start/home page" and then go down to the lower level the server will still be running? I went to the web site and looked at the documentation (at least some of it), it appears that the server will continue to run until I stop it. 

One last crazy one, do I need to be running Visual Studio Code in order to start my "start/home page" or can I just go the the "start/home page" and right lick to launch the server? I would like to have a way to run my set of web pages without having to use VSC to trigger the actions. I want to be able to show my web site to some others as though it is just a normal web site (again without showing them VSC.

BTW: I really appreciate this response.

**Reply by vehrmann** [reply]
> Are you then required to use Visual Studio Code to do your coding (not that I am not willing to give it a try)?

No, but I highly recommend to use it for coding. It has tons of features such as code highlighting etc.
You can also quickly update your code and view the result live (after reloading the page) 

> Does this mean that I can start Live Server and then run any pages I want to?
> I have a "set" of web pages and the problem I have been experiencing is a couple of layers down from the "start page". If I start with the "start/home page" and then go down to the lower level the server will still be running? 

Yes it will still be running.
Initially, you open a folder in VSC, run the live server, and you then view the index.html in your browser. If there is no index.html in this folder, a file directory will be displayed where you can chose files to be opened in the browser.
You can also open other pages than the index.html either directly or if linked within pages (your use case).

> One last crazy one, do I need to be running Visual Studio Code in order to start my "start/home page" or can I just go the the "start/home page" and right lick to launch the server? I would like to have a way to run my set of web pages without having to use VSC to trigger the actions. I want to be able to show my web site to some others as though it is just a normal web site (again without showing them VSC.

As far as i know, you need to open it via VSC. But you then can minimize VSC and the browser shows the page like a normal page on the web. Only the url seems odd (e.g. http://127.0.0.1:5000/your-folder/your-page.html), but you could hide the url bar when showing to others.
Just give it a try, it's free and easy to set up.

**Reply by ntxmodel** [reply]
Again, thanks for this feedback. I will have a look at it. If it works like I think, I could start my web page before showing it to my "viewer". I just want to see how it works as I drill down through the "layers" of my "site". 

I wish I could find a way to do this without having to launch the VSC (since I already have all of the code written for the pages I want to present).

**Reply by elalish** [reply]
I would recommend [Glitch](https://glitch.com/edit/#!/model-viewer) or other similar JS sandboxes. Did you remember to put in the `camera-controls` attribute?

**Reply by ntxmodel** [reply]
Yes, I had camera-controls in the model.

**Reply by ntxmodel** [reply]
Giving up. :(

I like the Visual Studio Code editor, but I am going to uninstall the Live Server extension. It did not work for me.

**Reply by ntxmodel** [reply]
Can anyone suggest/recommend a web site where I can try showing my glb files on a web page? I have tried many things suggested in this forum thread and from suggestions on the ThreeJS forum but cannot not get anything to work (show my model is AR - pan, rotate, and zoom). I do not have a web site and I only have my laptop computer. The original post on this thread shows (I hope) what I want to accomplish. I have several models (glb files) of car wash equipment that I would like to present in AR mode on a set of web pages. I have shared one of my files with others and they were able to view them fine. When I try to do it on my laptop (as the topic says), it AR mode does not work. I do not have a web site and only have my laptop, so I do not have a web server.

Again, can anyone give me some idea of where I might be able to try my web page to make sure I am adding the right code of how I could get a server to work on my laptop? Are there any "free web site hosts" where I could try?

**Reply by ntxmodel** [reply]
Since my model is a glb file, I cannot upload it on this forum.

**Reply by elalish** [reply]
What is AR on a laptop?

**Comment by ntxmodel** [comment]
In this case, it means the ability to pan, rotate and zoom. 
    On Monday, June 17, 2024 at 12:45:39 PM CDT, Emmett Lalish ***@***.***> wrote:   

 


What is AR on a laptop?

**Reply by elalish** [reply]
We call that 3D around here - AR (Augmented Reality) refers to 3D rendered on top of a camera feed, locked in to the physical surroundings. 

Anyway, this is literally what MV does by default with two lines of HTML. Can you please use the Glitch link I gave you above to show the problem you're having?

**Reply by ntxmodel** [reply]
I have things working fine now thanks to [TinyGLB](https://tinyglb.com/). I think the bottom line was because I do not have a web server running on my laptop and TinyGLB provides that service (along with some others) and a web server is needed to run model-viewer and its AR mode.
